### PR TITLE
Fix nvcc-imposed UB in `constexpr` functions

### PR DESCRIPTION
**Pull request by vuule**
## Description
nvcc does not support `constexpr` functions that are not well-defined to call from the device. This is UB even when the function is not called from the device.

Throwing an exception is one such operation. This PR cleans up error handling for functions that are called from device, and removes `constexpr` from the ones that are not actually used from the device, or in the constexpr context.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by karthikeyann — APPROVED** [review]

**Review by mythrocks — APPROVED** [review]
Wow, that is some unexpected weirdness.

**Comment by davidwendt** [comment]
Perhaps this is related to https://github.com/rapidsai/cudf/issues/7795

**Review by vyasr — APPROVED** [review]
I also suspect https://github.com/rapidsai/cudf/issues/7795 and mentioned that on Slack :)

**Review by bdice — APPROVED** [review]

**Comment by vuule** [comment]
There's definitely some overlap with https://github.com/rapidsai/cudf/issues/7795, but, AFAIK, that issue is related to device functions that include a throw. `RLE_stream_size` is only called on the host side. IMO it should behave according to the C++ standard.
Unless I misunderstand how code is generated and https://github.com/rapidsai/cudf/issues/7795 impacts both host and device code.

**Review by hyperbolic2346 — APPROVED** [review]

**Review by PointKernel — APPROVED** [review]

**Review by davidwendt — APPROVED** [review]

**Review by mythrocks — APPROVED** [review]

File: cpp/include/cudf/detail/utilities/device_operators.cuh [hunk 1]
```diff
@@ -83,7 +83,11 @@ struct DeviceSum {
   template <typename T, std::enable_if_t<cudf::is_fixed_point<T>()>* = nullptr>
   static constexpr T identity()
   {
+#ifndef __CUDA_ARCH__
```
Comment (mythrocks) [inline cpp/include/cudf/detail/utilities/device_operators.cuh:86]: Ah, shoot.  The Copyright year needs updating.

Comment (mythrocks) [reply cpp/include/cudf/detail/utilities/device_operators.cuh:86]: (I've made that tiny update.  Let's see if CI goes through now.)

**Review by mhaseeb123 — APPROVED** [review]

**Pull request merged**